homeassistant/components/anomaly/binary_sensor.py

```diff
@@ -0,0 +1,326 @@
+"""A sensor that detects anomalies in other components."""
+from collections import deque
+import logging
+from typing import Optional
+
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASSES_SCHEMA,
+    ENTITY_ID_FORMAT,
+    PLATFORM_SCHEMA,
+    BinarySensorEntity,
+)
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_FRIENDLY_NAME,
+    CONF_DEVICE_CLASS,
+    CONF_ENTITY_ID,
+    CONF_FRIENDLY_NAME,
+    CONF_SENSORS,
+    STATE_UNAVAILABLE,
+    STATE_UNKNOWN,
+)
+from homeassistant.core import State, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import generate_entity_id
+from homeassistant.helpers.event import async_track_state_change_event
+from homeassistant.helpers.reload import setup_reload_service
+from homeassistant.util import utcnow
+
+from . import DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_ATTRIBUTE = "attribute"
+ATTR_INVERT = "invert"
+ATTR_REQUIRE_BOTH = "require_both"
+ATTR_POSITIVE_ONLY = "positive_only"
+ATTR_NEGATIVE_ONLY = "negative_only"
+ATTR_MIN_CHANGE_AMOUNT = "min_change_amount"
+ATTR_MIN_CHANGE_PERCENT = "min_change_percent"
+ATTR_SAMPLE_DURATION = "sample_duration"
+ATTR_MAX_SAMPLES = "max_samples"
+ATTR_TRAILING_SAMPLE_DURATION = "trailing_sample_duration"
+ATTR_MAX_TRAILING_SAMPLES = "max_trailing_samples"
+ATTR_SAMPLE_COUNT = "sample_count"
+ATTR_TRAILING_SAMPLE_COUNT = "trailing_sample_count"
+ATTR_SAMPLE_AVG = "sample_average"
+ATTR_TRAILING_AVG = "trailing_sample_average"
+ATTR_CHANGE_AMOUNT = "change_amount"
+ATTR_CHANGE_PERCENT = "change_percent"
+
+CONF_ATTRIBUTE = "attribute"
+CONF_INVERT = "invert"
+CONF_POSITIVE_ONLY = "positive_only"
+CONF_NEGATIVE_ONLY = "negative_only"
+CONF_REQUIRE_BOTH = "require_both"
+CONF_MAX_SAMPLES = "max_samples"
+CONF_SAMPLE_DURATION = "sample_duration"
+CONF_MAX_TRAILING_SAMPLES = "max_trailing_samples"
+CONF_TRAILING_SAMPLE_DURATION = "trailing_sample_duration"
+CONF_MIN_CHANGE_AMOUNT = "min_change_amount"
+CONF_MIN_CHANGE_PERCENT = "min_change_percent"
+
+SENSOR_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Optional(CONF_ATTRIBUTE): cv.string,
+        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
```

Comment:
Why should we allow to configure the device class? We know what this sensor represents.

Reply:
Can there not be instances where the underlying device doesn't have a device_class? Granted they should fix that, but should we force that?

The base reason I had it is it is an option in the trend sensor, and I based this on the trend sensor.

Reply:
We're not representing the underlying device. We're representing an anomaly in a measurement series.

Reply:
This anomaly sensor is a binary sensor and I'm assuming the input data will come from a non binary sensor. So we can't use the input sensor device class to represent the anomaly binary sensor as these are different things.

Reply:
Fixed this.